### DISCUSSION TITLE
config option - for base url / subdir

### DISCUSSION THREAD

**Discussion by kcris** [photoview/photoview#507] [Q&A and Troubleshooting] [answered]
I am selfhosting multiple web apps, all of them offer some config option which allow me to specify the `base url`.
That base url is then used by those apps internally, as a prefix for all their internal urls.

This allows me to use a subdirectory on my webserver, for example I would like to expose photoview under `/photos` folder on my webserver, e.g `https://my.server.com/photos`

@viktorstrate : Is this possible with photoview? If not, could we have such an option? 

ps: I saw this thread https://github.com/photoview/photoview/discussions/392
but did not find a proper solution to this.

thanks

**Comment by viktorstrate** [comment, marked as answer]
This is possible but unfortunately only by building the Docker image yourself. The problem is that the UI is static files, and I don't want to tie it to the API as I want it to be as separate from the client as possible.

But by setting `UI_PUBLIC_URL=/photos` when building the image, that information will be baked into the static UI files.

https://github.com/photoview/photoview/blob/5eb4bf83f3fb923975683a5f1664b2350fcedfdf/Dockerfile#L32

You could do this by cloning the repo and changing the docker-compose file to this:

```diff
services:
  photoview:
-    image: viktorstrate/photoview:2
+    build:
+      context: .
+      args:
+        - UI_PUBLIC_URL=/photos
```

**Reply by yukdumboobumm** [reply]
I'd also like to see a CLI solution to this problem but in the meantime, if it helps anyone out, using a subdomain (ie gallery.example.com) and then proxying requests to the root of the subdomain will work just fine. Example for apache below:

```
<VirtualHost *:443>
        Protocols h2 http/1.1
        ServerName gallery.example.com

        DocumentRoot /var/www/example.com
        ErrorLog ${APACHE_LOG_DIR}/example.com/error.log

        Include /etc/letsencrypt/options-ssl-apache.conf
        SSLCertificateFile /etc/letsencrypt/live/example.com/fullchain.pem
        SSLCertificateKeyFile /etc/letsencrypt/live/example.com/privkey.pem

        SSLProxyEngine On
        RewriteEngine On
        RewriteCond %{HTTP:UPGRADE} ^WebSocket$ [NC,OR]
        RewriteCond %{HTTP:CONNECTION} ^Upgrade$ [NC]
        RewriteRule .* ws://127.0.0.1:8000%{REQUEST_URI} [P,QSA,L]

        RequestHeader set X-Forwarded-Proto expr=%{REQUEST_SCHEME}

        <Location />
                Require all granted
                ProxyPass http://127.0.0.1:8000/
                ProxyPassReverse http://127.0.0.1:8000/
                ProxyPassReverseCookieDomain 127.0.0.1 gallery.example.com
        </Location>

</VirtualHost>
````

edited to add the rewrite conditions for graphql websocket

**Reply by LuTzeee** [reply]
I struggle in getting Photoview running behind an **NGINX reverse-proxy**. Several other services are running fine, but photoview just delivers an apparently empty site. When looking closely, the `index.html` file seems to be delivered as expected, but the `index***.js` has the length of 0 and `index***.css`, `logo192.png` and `photoview-logo.svg` are the same as the `index.html` file. And no matter, which file under the (reverse-proxied) sub-domain "/fotos" I try to get, always the `index.html` is delivered.

I followed your advice and re-built the docker image with `UI_PUBLIC_URL` set. (Tried full URL with https://.../fotos and just /fotos).

My nginx snippet currently looks like this: 
```conf
location /fotos {
            # don't fail nginx when service is not up yet
            set $foto_upstream http://photoview; 
            proxy_pass $foto_upstream;
            proxy_set_header X-Forwarded-For $remote_addr;
            proxy_set_header Host $host;
            proxy_http_version 1.1;    # to keep alive
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection "upgrade";
}
```
Running without NGINX and without configured `UI_PUBLIC_URL` works fine. I would like to stick to NGINX, as this works fine for the rest.

If you have any suggestion, I would be happy to try :slightly_smiling_face:

**Reply by kkovaletp** [reply]
1. the original suggestion is from 2021, so I'd be surprised if it still works))
2. I didn't test it and don't have an Nginx config for this use case, but I assume that it might work if you copy the content of the `/app/ui` from the `photoview` container to the Nginx, set it to serve the frontend's static files by itself from the copied location, while redirecting only the `/api` requests to the `photoview` service's backend, and finally set the `PHOTOVIEW_SERVE_UI` for the `photoview` service to `false`

**Reply by LuTzeee** [reply]
Thank you so much for your suggestion. It got better - some of the static files are delivered, but I guess the dynamically generated ones are looked for at the wrong place (in my case, this file is not found: ` /usr/share/nginx/photoview_static/fotos/service-worker.js`. I assume, this should be retrieved via API-endpoint. )

I will try more but have no access to the machine for the next weeks - just wanted to give an intermediate feedback.

**Reply by kkovaletp** [reply]
There is an open issue for the `service-worker.js`: https://github.com/photoview/photoview/issues/938
It is also a static file, but for some reason the frontend builder doesn't even try to build it during the image creation. We need some help with the investigation of this issue ;-)

**Reply by LuTzeee** [reply]
I see ;-)
I will not be able to spend time on this the next 3 weeks and the used technologies are new to me. So I can't promise anything but I'll try my best and *maybe* report back in about a month.
Again, thank you for your free support :-)

**Reply by kkovaletp** [reply]
@LuTzeee, If the trick with the extraction of `/app/ui` from the `photoview` image and copying to the `nginx` work for you better, it would be useful for others if you could share the final Nginx config and the configuration steps to make everything work together

**Comment by kcris** [comment]
Seeing the previous [discussion](https://github.com/photoview/photoview/discussions/392#discussioncomment-831733) I was under the impression that 
* there are some invalid links even if using `UI_PUBLIC_URL` as docker arg 
* I have to disable the embedded web server by setting `PHOTOVIEW_SERVE_UI=0` and using my own webserver to serve photoview

@viktorstrate Is that the case?

I will give it a try, as long as it works, it's a perfect solution (does not have to be a runtime option, a docker arg will do just fine)
@viktorstrate thanks a lot

**Reply by oliverhanappi** [reply]
@kcris I am facing the same issue. Did you manage to get it running? Can you post your docker-compose file and proxy configuration?
